backend/src/routes/classroom-group.ts

```diff
@@ -4,11 +4,122 @@ import { sql } from '@/lib/db';
+    .use(middleware)
+    .ws('/:group_slug/g/chat', {
+        async open(ws) {
+            const {
+                user,
+                session,
+                params: { group_slug: groupSlug },
+            } = ws.data;
+
+            if (!user || !session) {
+                ws.send({
+                    status: 'error',
+                    message: 'Unauthenticated, Please sign in and try again',
+                });
+                ws.close();
+                return;
+            }
+
+            const chatHistory = await sql`
+            SELECT
+                content,
+                created_at,
+                created_by
+            FROM group_message
+            WHERE group_slug = ${groupSlug}
+            `;
+
+            const usernameRecords = await sql`
+            SELECT DISTINCT
+                auth_user.id,
+                auth_user.first_name || ' ' || auth_user.last_name AS "fullName"
+            FROM auth_user
+            INNER JOIN group_message
+                ON group_message.created_by = auth_user.id
+            `;
+
+            let usernames: {
+                [key: string]: string;
+            } = {};
+
+            // transfer username records to a dictionary
+            for (let i = 0; i < usernameRecords.length; i++) {
+                const key: string = usernameRecords[i].id;
+                usernames[key] = usernameRecords[i].fullName;
+            }
+
+            for (const message of chatHistory) {
+                ws.send({
+                    message: message.content,
+                    created_at: message.created_at,
+                    created_by: usernames[message.created_by],
+                });
+            }
+
+            ws.subscribe(groupSlug);
+        },
+        async message(ws, message) {
+            const {
+                user,
+                session,
+                params: { group_slug: groupSlug },
+            } = ws.data;
+
+            if (!user || !session) {
+                ws.send({
+                    status: 'error',
+                    message: 'Unauthenticated, Please sign in and try again',
+                });
+                ws.close();
+                return;
+            }
+
+            const usernameRecords = await sql`
+                SELECT DISTINCT
+                    auth_user.id,
+                    auth_user.first_name,
+                    auth_user.last_name
+                FROM auth_user INNER JOIN group_message
+                ON group_message.created_by = auth_user.id
+            `;
+
+            let usernames: {
+                [key: string]: string;
+            } = {};
+
+            await sql`
+                INSERT INTO group_message
+                    ( content, created_by, group_slug )
+                VALUES
+                    ( ${(message as { message: string; type: string }).message}, ${user.id}, ${groupSlug.toString()} )
+            `;
```

Comment:
If this is a await, this will slow message because it will wait for database to finish then publish a message